portia/tool_registry.py

```diff
@@ -192,6 +194,40 @@ def filter_tools(self, predicate: Callable[[Tool], bool]) -> ToolRegistry:
         """
         return ToolRegistry({tool.id: tool for tool in self._tools.values() if predicate(tool)})
 
+    def with_tool_description(
+        self, tool_id: str, updated_description: str, *, overwrite: bool = False
+    ) -> None:
+        """Update a tool with an extension or override of the tool description.
+
+        Args:
+            tool_id (str): The id of the tool to update.
+            updated_description (str): The tool description to update. If `overwrite` is False, this
+                will extend the existing tool description, otherwise, the entire tool description
+                will be updated.
+            overwrite (bool): Whether to update or extend the existing tool description.
+
+        Returns:
+            None: The tool registry is updated in place.
+
+        Particularly useful for customising tools in MCP servers for usecases. A deep copy is made
+        of the underlying tool such that the tool description is only updated within this registry.
+        Logs a warning if the tool is not found.
+
+        """
+        try:
+            tool = self.get_tool(tool_id=tool_id)
+            new_tool = tool.__deepcopy__()
+            new_description = (
+                updated_description
+                if overwrite
+                else f"{new_tool.description}. {updated_description}"
+            )
+            new_tool.description = new_description
```

Comment:
You could use Pydantic's `model_copy` here ([docs](https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_copy))

So this could be 

```suggestion
            new_tool = tool.model_copy(
                update={"description": new_description},
                deep=True
            )
```